x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -295,6 +318,23 @@ private void buildThenMaybeCacheRole(RoleKey roleKey, Collection<RoleDescriptor>
         }, listener::onFailure));
     }
 
+    private void getOrBuildRoleForApiKey(Authentication authentication, boolean limitedBy, ActionListener<Role> roleActionListener) {
+        final Tuple<String, BytesReference> apiKeyIdAndBytes = apiKeyService.getApiKeyIdAndRoleBytes(authentication, limitedBy);
```

Comment:
Methods here don't have a clear interface, it could just as well not be methods at all; the `limitedBy` flag is really ruining every structure.

I would move the full logic that obtains the `Role` for an API Keys from the `CompositeRolesStore` class to the `ApiKeyService` class.
But I realise I'm late to the party, we can do this in a follow-up.
@tvernum would you agree merging this PR, and following up with a refactoring?

Reply:
`Role` building has always been done inside `CompositeRolesStore` and I think it fits here since role building and caching are the main purposes of `CompositeRolesStore`. It only delegates retrieval of role descriptors to different roles providers, including `ApiKeyService`, which is the case for this method as well. Overall, I think the method name could use some help (maybe `buildAndCacheRoleForApiKey` to mirror the existing `buildAndCacheRoleFromDescriptors`?), but its content looks fine to me.

Reply:
Good point! Yes, `Role` should stay inside the `CompositeRolesStore`.

Then I think we can make `ApiKeyRoleDescriptors` lazily parse RoleDescriptor. Not sure what to do about the cache key. Let's leave this for the future us.